core/src/main/java/com/graphhopper/storage/CHGraphImpl.java

```diff
@@ -176,20 +155,23 @@ int weightToWeightFlags(long edgePointer, double weight) {
                     " nodeB " + nodeAccess.getLat(getNodeB(edgePointer)) + "," + nodeAccess.getLon(getNodeB(edgePointer)));
             weight = MIN_WEIGHT;
         }
-        if (weight > MAX_WEIGHT)
-            weightInt = MAX_WEIGHT_31;
+        if (weight >= WEIGHT_INFINITY)
+            weightInt = (int) WEIGHT_LONG_INFINITY; // negative
         else
-            weightInt = ((int) Math.round(weight * WEIGHT_FACTOR)) << 2;
-        return weightInt;
+            weightInt = (int) Math.round(weight * WEIGHT_FACTOR);
+
+        shortcuts.setInt(edgePointer + S_WEIGHT, weightInt);
     }
 
     double getShortcutWeight(long edgePointer) {
-        // no need for reverseFlags call (shortcut has identical weight if both dies) and also no need for 64bit
-        long flags32bit = getShortcutFlags(edgePointer);
-        double weight = (flags32bit >>> 2) / WEIGHT_FACTOR;
-        if (weight >= MAX_WEIGHT)
+        // might be negative after converting into long -> we need to remove this sign via "& 0xFFFFFFFF". The L is necessary or prepend 8 zeros.
```

Comment:
I really meant "into long" here. We have the int value from shortcuts DataAccess and when it is converted into a long value the JVM might preserve the sign and to remove this we apply `& 0xFFFFFFFFL`.

Reply:
Ah ok. I thought you mean the stored (int) value can be negative after we cast from long to int in `setWeight`. I would not expect the cast from int to long here to change the value at all?! So actually if the value is negative it is negative because the value we put in exceeded Integer.MAX_VALUE and the 'overflow' results in a negative value when casting to int?

Reply:
> and the 'overflow' results in a negative value when casting to int?

When casting to long, yes.

Let's say we have Integer.max+1, i.e. a negative value if stored as `int`. If we convert this `int` into a long it is actually the same negative value but needs to fill all higher bits with a 1 and we can now convert this negative value via 2's complement conversion or with `& 0xFFFFFFFF`.

Let's use some visualisation of the bits:

```
00000000 00000000 00000000 00000000 01111111 11111111 11111111 11111111 = Integer.toBinaryString(Integer.MAX_VALUE)
00000000 00000000 00000000 00000000 10000000 00000000 00000000 00000000 = Integer.toBinaryString(Integer.MAX_VALUE + 1) 
```

Now because `Integer.MAX_VALUE + 1` is negative and we cast it to `long` the JVM fills everything with a 1 to keep it the same negative value:

```
11111111 11111111 11111111 11111111 10000000 00000000 00000000 00000000 = Long.toBinaryString(Integer.MAX_VALUE + 1)
```

And to convert this 2's complement into the same positive value we could do `~val+1` or `val & 0xFFFFFFFFL`:
```
00000000 00000000 00000000 00000000 10000000 00000000 00000000 00000000 = Long.toBinaryString(((long) (Integer.MAX_VALUE + 1)) & 0xFFFFFFFFL) = Long.toBinaryString(~((long) (Integer.MAX_VALUE + 1)) + 1)
```

Reply:
> When casting to long, yes.

Hm, but the value becomes negative when casting from long to int, not when casting from int to long. 
I meant this:

```java
        // setShortcutWeight
        // Integer.MAX_VALUE = 2_147_483_648
        double weight = 2147483.649; // (Integer.MAX_VALUE+1)/1000.0
        System.out.println(weight); // 2147483.649
        double weightMultFac = weight * 1000; 
        System.out.println(weightMultFac); // 2147483649.0
        long rounded = Math.round(weightMultFac);
        System.out.println(rounded); // 2147483649
        int cast = (int) rounded;
        System.out.println(cast); // -2147483647, this is where the value becomes negative: when casting from long to int

        // getShortcutWeight
        long longcast = (long)cast; 
        System.out.println(longcast); // still -2147483647, casting to long doesn't change the sign, the (explicit long) cast is not even necessary
        long resultLong = longcast & 0xFFFFFFFFL;
        System.out.println(resultLong); // 2147483649
        double result = resultLong / 1000.0;
        System.out.println(result); // 2147483.649
        assertEquals(weight, result);
```

Reply:
Ah, yes, this is incorrectly phrased. Should we change it to:
```
If the value is too large (> Integer.MAX_VALUE) the `int` is negative. Converted to `long` the JVM fills the high bits with 1's which we remove via "& 0xFFFFFFFFL" to get the unsigned value. (The L is necessary or prepend 8 zeros.)
```